### ISSUE TITLE
Don't await SocketInput directly. Use a TaskQueue instead.

### ISSUE THREAD

**Issue by halter73** [aspnet/KestrelHttpServer#456]
Instead of simply throwing for attempted concurrent reads (i.e. two callers awaiting SocketInput and/or Request.Body.ReadAsync simultaneously), we could queue up reads in a thread-safe manner using something like the [TaskQueue](https://github.com/SignalR/SignalR/blob/3a219c101587333d562f04f30903d50ad9773e1c/src/Microsoft.AspNet.SignalR.Core/Infrastructure/TaskQueue.cs) in SignalR.

I personally don't support this change, because concurrent reads are indicative of an application error that should be surfaced to the user if possible. There is also a potential performance hit since every time we await, even if data is already available, we would be allocating a new Func. This includes awaits Kestrel does internally. Maybe there is a way to modify TaskQueue to avoid this though.

Nonetheless, this was suggested by @lodejard, so I'm creating this issue so we can discuss the idea.

**Comment by benaadams** [comment]
Don't like; the error was caused by an optimization which didn't translate well - so this would be a de-optimization; overall. Will add detail to the other issue.

**Comment by benaadams** [comment]
Actaully - let me dig into what was happening first

**Comment by benaadams** [comment]
Yeah disagree with this; its a bit like trying to do 2+ concurrent reads on a stream; if you can detect and error out as an option its a better approach than queuing and hiding the issue which would add a slow path for even non-concurrent reads.

To have multiple things wait on a single read a new task can always be constructed using an async method and `var newTask = Task.WhenAll(AwaitingMethod, CompletedTask)` which then can have multiple awaiters attached? But I don't see it being a common occurrence.
